consensus/reactor.go

```diff
@@ -762,967 +688,701 @@ func (conR *Reactor) gossipVotesForHeight(
+// execution will result in a PeerError being sent on the VoteSetBitsChannel.
+// When the reactor is stopped, we will catch the signal and close the p2p
+// Channel gracefully.
+func (r *Reactor) processVoteSetBitsCh() {
+	defer r.voteSetBitsCh.Close()
 
-// HasVoteMessage is sent to indicate that a particular vote has been received.
-type HasVoteMessage struct {
-	Height int64
-	Round  int32
-	Type   tmproto.SignedMsgType
-	Index  int32
-}
+	for {
+		select {
+		case envelope := <-r.voteSetBitsCh.In:
+			if err := r.handleMessage(r.voteSetBitsCh.ID, envelope); err != nil {
+				r.Logger.Error("failed to process message", "ch_id", r.voteSetBitsCh.ID, "envelope", envelope, "err", err)
+				r.voteSetBitsCh.Error <- p2p.PeerError{
+					NodeID: envelope.From,
+					Err:    err,
+				}
+			}
 
-// ValidateBasic performs basic validation.
-func (m *HasVoteMessage) ValidateBasic() error {
-	if m.Height < 0 {
-		return errors.New("negative Height")
-	}
-	if m.Round < 0 {
-		return errors.New("negative Round")
-	}
-	if !types.IsVoteTypeValid(m.Type) {
-		return errors.New("invalid Type")
-	}
-	if m.Index < 0 {
-		return errors.New("negative Index")
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on VoteSetBitsChannel; closing...")
+			return
+		}
 	}
-	return nil
 }
 
-// String returns a string representation.
-func (m *HasVoteMessage) String() string {
-	return fmt.Sprintf("[HasVote VI:%v V:{%v/%02d/%v}]", m.Index, m.Height, m.Round, m.Type)
-}
-
-//-------------------------------------
+// processPeerUpdates initiates a blocking process where we listen for and handle
+// PeerUpdate messages. When the reactor is stopped, we will catch the signal and
+// close the p2p PeerUpdatesCh gracefully.
+func (r *Reactor) processPeerUpdates() {
+	defer r.peerUpdates.Close()
 
-// VoteSetMaj23Message is sent to indicate that a given BlockID has seen +2/3 votes.
-type VoteSetMaj23Message struct {
-	Height  int64
-	Round   int32
-	Type    tmproto.SignedMsgType
-	BlockID types.BlockID
-}
+	for {
+		select {
+		case peerUpdate := <-r.peerUpdates.Updates():
+			r.processPeerUpdate(peerUpdate)
 
-// ValidateBasic performs basic validation.
-func (m *VoteSetMaj23Message) ValidateBasic() error {
-	if m.Height < 0 {
-		return errors.New("negative Height")
-	}
-	if m.Round < 0 {
-		return errors.New("negative Round")
-	}
-	if !types.IsVoteTypeValid(m.Type) {
-		return errors.New("invalid Type")
-	}
-	if err := m.BlockID.ValidateBasic(); err != nil {
-		return fmt.Errorf("wrong BlockID: %v", err)
+		case <-r.closeCh:
+			r.Logger.Debug("stopped listening on peer updates channel; closing...")
+			return
+		}
 	}
-	return nil
 }
 
-// String returns a string representation.
-func (m *VoteSetMaj23Message) String() string {
-	return fmt.Sprintf("[VSM23 %v/%02d/%v %v]", m.Height, m.Round, m.Type, m.BlockID)
-}
+func (r *Reactor) peerStatsRoutine() {
+	for {
+		if !r.IsRunning() {
+			r.Logger.Info("stopping peerStatsRoutine")
+			return
```

Comment:
should we flush `statsMsgQueue` before exiting?

Reply:
We can execute `close(statsMsgQueue)` during `State#OnStop`, so it's closed when the reactor closes. WDYT?

Reply:
Then again, the `State` type doesn't close any of its channels. Not sure if its worth it?

Reply:
I don't think we should close statsMsgQueue - will keep us away from panicking. You mean old implementation does not do flushing? Then I guess it's fine. But this is what I would've done to ensure all goroutines that are writing to statsMsgQueue terminate.

Reply:
> You mean old implementation does not do flushing? 

Yes, correct. In fact, `peerStatsRoutine` was largely left untouched.